core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/WorkflowScheduler.scala

```diff
@@ -0,0 +1,394 @@
+        if (!builtOperators.contains(op)) {
+          buildOperator(prev, op)
+          builtOperators.add(op)
+        }
+        builtOpsInRegion.add(op)
+      }
+
+      frontier = (region
+        .getOperators() ++ region.blockingDowstreamOperatorsInOtherRegions)
+        .filter(opId => {
+          !builtOpsInRegion.contains(opId) && workflow
+            .getDirectUpstreamOperators(opId)
+            .filter(region.getOperators().contains)
+            .forall(builtOperators.contains)
+        })
+    }
+  }
+
+  private def buildOperator(
+      prev: Array[(OperatorIdentity, WorkerLayer)], // used to decide deployment of workers
+      operatorIdentity: OperatorIdentity
+  ): Unit = {
+    val opExecConfig = workflow.getOperator(operatorIdentity)
+    if (opExecConfig.topology.links.isEmpty) {
+      opExecConfig.topology.layers.foreach(workerLayer => {
+        workerLayer.build(
+          prev.map(pair => (workflow.getOperator(pair._1), pair._2)),
+          availableNodes,
+          networkCommunicationActor.ref,
+          ctx,
+          workflow.getInlinksIdsToWorkerLayer(workerLayer.id),
+          workflow.workerToLayer,
+          workflow.workerToOperatorExec
+        )
+      })
+    } else {
+      val operatorInLinks: Map[WorkerLayer, Set[WorkerLayer]] =
+        opExecConfig.topology.links.groupBy(_.to).map(x => (x._1, x._2.map(_.from).toSet))
+      var layers: Iterable[WorkerLayer] =
+        opExecConfig.topology.links
+          .filter(linkStrategy => opExecConfig.topology.links.forall(_.to != linkStrategy.from))
+          .map(_.from) // the first layers in topological order in the operator
+      layers.foreach(workerLayer => {
+        workerLayer.build(
+          prev.map(pair => (workflow.getOperator(pair._1), pair._2)),
+          availableNodes,
+          networkCommunicationActor.ref,
+          ctx,
+          workflow.getInlinksIdsToWorkerLayer(workerLayer.id),
+          workflow.workerToLayer,
+          workflow.workerToOperatorExec
+        )
+      })
+      layers = operatorInLinks.filter(x => x._2.forall(_.isBuilt)).keys
+      while (layers.nonEmpty) {
+        layers.foreach((layer: WorkerLayer) => {
+          layer.build(
+            operatorInLinks(layer).map(y => (null, y)).toArray,
+            availableNodes,
+            networkCommunicationActor.ref,
+            ctx,
+            workflow.getInlinksIdsToWorkerLayer(layer.id),
+            workflow.workerToLayer,
+            workflow.workerToOperatorExec
+          )
+        })
+        layers = operatorInLinks.filter(x => !x._1.isBuilt && x._2.forall(_.isBuilt)).keys
+      }
+    }
+  }
+
+  private def prepareRegion(region: PipelinedRegion): Future[Unit] = {
+    val allOperatorsInRegion =
+      region.getOperators() ++ region.blockingDowstreamOperatorsInOtherRegions
+
+    Future(asyncRPCClient.sendToClient(WorkflowStatusUpdate(workflow.getWorkflowStatus)))
```

Comment:
`sendToClient` doesn't need to be wrapped in a future, 
in `AsyncRPCClient`, in `sentToClient`, you can see `sendToClient` just returns a `Unit` instead of `Future[Unit]`,

This is because `sendToClient` is just fire and forget, and it doesn't expect a response. We can use it just like a normal function call. Even if we wrap it in a `Future`, this future would just return immediately.

There is another use of `sendToClient` down below, similary comments apply.

Reply:
done